Update the README and the landing page (#1095)
holoviz/param: README.md

processor = Processor(verbose=True)

# A Parameterized instance behaves as a usual instance of a class.
processor.run()
# => [fast] retry=3

# Parameterized implements a nice and simple repr.
print(repr(processor))
# => Processor(mode='accurate', name='Processor00042', retries=3, verbose=False)

# Attempting to set the retries attribute to 42 will raise
# an error as the value must be between 0 and 10.
try:
    processor.retries = 42
except ValueError as e:
    print(e)
    # => Integer parameter 'Processor.retries' must be at most 10, not 42.

# The `.param` namespace allows to get a hold on Parameter objects via
# indexing or attribute access.
print(processor.param['mode'].objects)  # => ['fast', 'accurate']
print(processor.param.mode.objects)  # => ['fast', 'accurate']

# This namespace offers many more useful methods, such as `.update()`
# to update multiple parameters at once, or `.values()` to obtain
# a dict of parameter name to parameter value.
processor.param.update(mode='accurate', verbose=False)
print(processor.param.values())
# => {'mode': 'accurate', 'name': 'Processor00042', 'retries': 3, 'verbose': False}
```

> Runtime attribute validation is a great feature that helps build defendable code bases! Alternative libraries, like [Pydantic](https://docs.pydantic.dev) and others, excel at input validation, and if this is only what you need, you should probably look into them. Where Param shines is when you also need:

Author: Yeah, I think they've gone back and forth on how they want to describe themselves between validation and parsing/coercion. In v2, they added a strict mode (configurable per class or per field) that prevents coercion, and decided to stick to "validation" (more info in the middle of this long blog post https://pydantic.dev/articles/pydantic-v2#im-dropping-the-word-parse-and-just-using-validate).

<img width="1355" height="516" alt="image" src="https://github.com/user-attachments/assets/459e93af-d852-4f24-8c7d-34444d4ff0e9" />

I'd prefer to stick to "validation" here as a general term.

We'll start with APIs that trigger side-effects only, which either have the name watch in their name or are invoked with `watch=True`:

1. `<parameterized_obj>.param.watch(fn, *parameters, ...)`: Low-level, imperative API to attach callbacks to parameter changes, the callback receives one or more rich `Event` objects.
2. `@depends(*parameter_names, watch=True)`: In a Parameterized class, declare dependencies and automatically watch parameters for changes to call the decorated method.
3. `@depends(*parameters, watch=True)`: Equivalent of 2 above to decorate functions.
4. `bind(fn, *references, watch=True, **kwargs)`: Function binding with automatic references (parameters, bound functions, reactive expressions) watching and triggering on changes.

Running the snippet below, you will see that updating the parameters `a`, `b`, `c`, and `d` will emit print statements.

```python
import param

class SideEffectExample(param.Parameterized):
    a = param.String()
    b = param.String()
    c = param.String()
    d = param.String()

    def __init__(self, **params):
        super().__init__(**params)
        self.param.watch(lambda event: print(event), 'a')  # 1.

> 

Author: Well, I could remove it, instantiate the class, and then attach the watcher with `sfe.param.watch(...)` but it's not the same, as what I generally want is for all the instances to have the watcher set up. I believe that having to override `__init__` is very common when writing Parameterized classes (and is a reason why I want to add `post_init`).

    @param.depends('b', watch=True)  # 2.
    def print_b(self):
        print(f"print_b: {self.b=}")

sfe = SideEffectExample()
sfe.a = 'foo'
# Event(what='value', name='a', obj=SideEffectExample(a='foo', b='', c='', d='', name='SideEffectExample00015'), cls=Example(a='foo', b='', c='', d='', name='SideEffectExample00015'), old='', new='foo', type='changed')

sfe.b = 'bar'
# print_b: self.b='bar'

> @param.depends(sfe.param.c, watch=True)  # 3.

Reviewer: I would suggest not documenting using `depends()` outside of a class because of how the decorator syntax encourages early binding to a reactive object, which I think limits how the function can be used. I don't know of any advantage that usage has over `bind` and would always steer users to `bind` instead when using functions.

def print_c(c):
    print(f"print_c: {c=}")

sfe.c = 'baz'
# print_c: c='baz'

def print_d(d):
    print(f"print_d: {d=}")

param.bind(print_d, sfe.param.d, watch=True)  # 4.

sfe.d = 'biz'
# print_d: d='biz'
```

Let's continue the tour with what we'll call "reactive APIs". Contrary to the APIs presented above, in this group parameter updates do not immediately trigger side effects. Instead, these APIs let you declare relationships, dependencies, and expressions, which can be introspected by other frameworks to set up their own reactive workflows.

1. `@depends(*parameter_names)`: In a Parameterized class, declare parameter dependencies by decorating a method.
2. `@depends(*parameters)`: Equivalent of 1 above to decorate functions.
3. `bind(fn, *references, **kwargs)`: Create a bound function, that when called, will always use the current parameter/reference value. `bind` is in a way a reactive version of [`functools.partial`](https://docs.python.org/3/library/functools.html#functools.partial).

> 4. `rx()`: Fluent API to create reactive expressions, that allows chaining and composing operations.

Reviewer: ```suggestion
4. `rx()`: Fluent API to create reactive expressions, which allow chaining and composing operations.
```

```python
import param

class ReactiveExample(param.Parameterized):
    x = param.Integer()
    y = param.Integer()

    @param.depends('x', 'y')  # 1.
    def sum(self):
        return self.x + self.y

re = ReactiveExample()

@param.depends(re.param.x, re.param.y)  # 2.
def sub(a, b):
    return a - b

def mul(a, b):
    return a * b

bound_mul = param.bind(mul, re.param.x, re.param.y)  # 3.
re.param.update(x=2, y=4)
bound_mul()
# 8
```

`rx()` is the highest-level reactive API Param offers. We'll show a simple example first, using literal values as input of three source reactive expressions, that we combine with simple arithmetic operations.

```python
from param import rx

> # Reactive expressions can be created by simply calling an object with `rx()`

Reviewer: I don't think it's valid to describe `rx(y)` as "calling y". It's calling `rx` _with_ `y`, right?

```suggestion
# Reactive expressions can be created by simply wrapping an object with `rx()`
```

# `val2`, and then multiplying the sum with `factor`.

# We can obtain the resolved value of `res` via the `.rx.value` property
# (`.rx` being a namespace holding other useful attributes and methods).
print(res.rx.value)  # => 0, from (0 + 0) * 1

# We can then update one of the source reactive expressions by setting
# it via the `.rx.value` property.
val1.rx.value = 2
# Requesting the value of `res` resolves its updated value (2 instead of 0).
print(res.rx.value)  # => 2, from (2 + 0) * 1

# Similarly, we can update `factor` and request the new value of `res`.
# Note that only the last multiplication is recomputed; the intermediate
# result of `val1 + val2` is automatically cached.
factor.rx.value = 10
print(res.rx.value)  # => 20, from (2 + 0) * 10
```

The snippet below shows a slightly more complex example that reveals the power of reactive expressions.

```python
import param

class RXExample(param.Parameterized):
    val1 = param.String('foo')
    val2 = param.String('bar')

example = RXExample()

# Reactive expressions can be created directly from Parameters by
# calling `.rx()`. The source of the reactive expression is not a

> # literal value, but the more dynamic Parameter object.

Reviewer: ```suggestion
# literal value, but a fully dynamic Parameter object.
```
